File: commons-csv-python/src/main/python/csv_printer.py
from main.python.closeable import Closeable
from main.python.java_handler import Types, java_handler

@java_handler
class CSVPrinter(Closeable):
    CR = "\r"
    LF = "\n"
    SP = " "

    def __init__(self, out, format):
        try:
            assert out is not None
            assert format is not None
        except:
            raise ValueError("out and format must not be None")

        self.out = out
        self.format = format
        self.new_record = True
        if format.get_header_comments():
            for line in format.get_header_comments():
                if line is not None:
                    self.print_comment(line)
        if format.get_header() and not format.get_skip_header_record():
            self.print_record(*format.get_header())

    def close(self, flush=False):
        if flush or self.format.get_auto_flush():
            if hasattr(self.out, "flush"):
                self.out.flush()
        if hasattr(self.out, "close"):
            self.out.close()

    def flush(self):
        if hasattr(self.out, "flush"):
            self.out.flush()

    def get_out(self):
        return self.out

    def print(self, value):
        self.format.print(value, self.out, self.new_record)
        self.new_record = False

    def _append(self, appendable, string):
        if hasattr(appendable, "append"):
            try: appendable.append(string)
            except: appendable.append(string.encode("utf-8"))
        else:
            try: appendable.write(string)
            except: appendable.write(string.encode("utf-8"))

    def print_comment(self, comment):
        if not self.format.is_comment_marker_set():
            return
        if not self.new_record:
            self.println()
        self._append(self.out, self.format.get_comment_marker())
        self._append(self.out, self.SP)
        for i in range(len(comment)):
            c = comment[i]
            if c == self.CR:
                if i + 1 < len(comment) and comment[i + 1] == self.LF:
                    i += 1
                self.println()
                self._append(self.out, self.format.get_comment_marker())
                self._append(self.out, self.SP)
            elif c == self.LF:
                self.println()
                self._append(self.out, self.format.get_comment_marker())
                self._append(self.out, self.SP)
            else:
                self._append(self.out, c)
        self.println()

    def println(self):
        self.format.println(self.out)
        self.new_record = True

    def __check_all_strings(self, *values):
        if values is None:
            return False

        for value in values:
            if not isinstance(value, str):
                return False
        return True

    def print_record(self, *values):        
        if len(values) == 1 and \
                hasattr(values[0], "__iter__") and \
                not isinstance(values[0], str): # check if values[0] is iterable but is not a string!
            for value in values[0]:
                self.print(value)
            self.println()
        else:
            self.format.print_record(self.out, *values)
            self.new_record = True

    def print_records(self, *values):      
        if len(values) == 1 and isinstance(values[0], Types.Cursor):
            return self.print_records_result_set(values[0])
        
        for value in values:            
            value = value if value != None else []
            
            # check if value is a list of records
            if self.__check_all_strings(*value):
                self.print_record(*value)
            else:
                # value is a list of lists/iterables
                for record in value:
                    formatted_value = record if record != None else []
                    self.print_record(*formatted_value)

    def print_records_result_set(self, result_set):
        # result_set is java.sql.ResultSet
        if hasattr(result_set, "next"):
            column_count = result_set.getMetaData().getColumnCount()
            while result_set.next():
                for i in range(1, column_count+1):
                    self.print(result_set.getObject(i))
                self.println()
        # result_set is python iterable
        else:
            for line in result_set:
                for e in line:
                    self.print(e)
                self.println()

File: commons-csv-graal/src/main/java/org/apache/commons/csv/CSVPrinter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.commons.csv;

import static org.apache.commons.csv.Constants.CR;
import static org.apache.commons.csv.Constants.LF;
import static org.apache.commons.csv.Constants.SP;

import java.io.Closeable;
import java.io.Flushable;
import java.io.IOException;
import java.sql.ResultSet;
import java.sql.SQLException;

import org.graalvm.polyglot.Value;

/**
 * Prints values in a CSV format.
 */
public final class CSVPrinter implements Flushable, Closeable {

    /** glue code attributes */
    private static Value clz = ContextInitializer.getPythonClass("csv_printer.py", "CSVPrinter"); 
    private Value obj;

    /**
     * Creates a printer that will print values to the given stream following the CSVFormat.
     * <p>
     * Currently, only a pure encapsulation format or a pure escaping format is supported. Hybrid formats (encapsulation
     * and escaping with a different character) are not supported.
     * </p>
     *
     * @param out
     *            stream to which to print. Must not be null.
     * @param format
     *            the CSV format. Must not be null.
     * @throws IOException
     *             thrown if the optional header cannot be printed.
     * @throws IllegalArgumentException
     *             thrown if the parameters of the format are inconsistent or if either out or format are null.
     */
    public CSVPrinter(final Appendable out, final CSVFormat format) throws IOException {
        Assertions.notNull(out, "out");
        Assertions.notNull(format, "format");

        obj = clz.newInstance(out, format);
    }

    /**
     * @return the Python guest CSVPrinter associated with this printer.
     */
    public Value getPythonObject() {
        return obj;
    }

    // ======================================================
    // printing implementation
    // ======================================================

    @Override
    public void close() throws IOException {
        close(false);
    }

    /**
     * Closes the underlying stream with an optional flush first.
     * @param flush whether to flush before the actual close.
     *
     * @throws IOException
     *             If an I/O error occurs
     * @since 1.6
     */
    public void close(final boolean flush) throws IOException {
        obj.invokeMember("close", flush);
    }

    /**
     * Flushes the underlying stream.
     *
     * @throws IOException
     *             If an I/O error occurs
     */
    @Override
    public void flush() throws IOException {
        obj.invokeMember("flush");
    }

    /**
     * Gets the target Appendable.
     *
     * @return the target Appendable.
     */
    public Appendable getOut() {
        return obj.invokeMember("get_out").as(Appendable.class);
    }

    /**
     * Prints the string as the next value on the line. The value will be escaped or encapsulated as needed.
     *
     * @param value
     *            value to be output.
     * @throws IOException
     *             If an I/O error occurs
     */
    public void print(final Object value) throws IOException {
        obj.invokeMember("print", value);
    }

    /**
     * Prints a comment on a new line among the delimiter separated values.
     *
     * <p>
     * Comments will always begin on a new line and occupy a least one full line. The character specified to start
     * comments and a space will be inserted at the beginning of each new line in the comment.
     * </p>
     *
     * If comments are disabled in the current CSV format this method does nothing.
     *
     * @param comment
     *            the comment to output
     * @throws IOException
     *             If an I/O error occurs
     */
    public void printComment(final String comment) throws IOException {
        obj.invokeMember("print_comment", comment);
    }

    /**
     * Outputs the record separator.
     *
     * @throws IOException
     *             If an I/O error occurs
     */
    public void println() throws IOException {
        obj.invokeMember("println");
    }

    /**
     * Prints the given values a single record of delimiter separated values followed by the record separator.
     *
     * <p>
     * The values will be quoted if needed. Quotes and newLine characters will be escaped. This method adds the record
     * separator to the output after printing the record, so there is no need to call {@link #println()}.
     * </p>
     *
     * @param values
     *            values to output.
     * @throws IOException
     *             If an I/O error occurs
     */
    public void printRecord(final Iterable<?> values) throws IOException {
        obj.invokeMember("print_record", values);
    }

    /**
     * Prints the given values a single record of delimiter separated values followed by the record separator.
     *
     * <p>
     * The values will be quoted if needed. Quotes and newLine characters will be escaped. This method adds the record
     * separator to the output after printing the record, so there is no need to call {@link #println()}.
     * </p>
     *
     * @param values
     *            values to output.
     * @throws IOException
     *             If an I/O error occurs
     */
    public void printRecord(final Object... values) throws IOException {
        obj.invokeMember("print_record", values);
    }

    /**
     * Prints all the objects in the given collection handling nested collections/arrays as records.
     *
     * <p>
     * If the given collection only contains simple objects, this method will print a single record like
     * {@link #printRecord(Iterable)}. If the given collections contains nested collections/arrays those nested elements
     * will each be printed as records using {@link #printRecord(Object...)}.
     * </p>
     *
     * <p>
     * Given the following data structure:
     * </p>
     *
     * <pre>
     * <code>
     * List&lt;String[]&gt; data = ...
     * data.add(new String[]{ "A", "B", "C" });
     * data.add(new String[]{ "1", "2", "3" });
     * data.add(new String[]{ "A1", "B2", "C3" });
     * </code>
     * </pre>
     *
     * <p>
     * Calling this method will print:
     * </p>
     *
     * <pre>
     * <code>
     * A, B, C
     * 1, 2, 3
     * A1, B2, C3
     * </code>
     * </pre>
     *
     * @param values
     *            the values to print.
     * @throws IOException
     *             If an I/O error occurs
     */
    public void printRecords(final Iterable<?> values) throws IOException {
        obj.invokeMember("print_records", values);
    }

    /**
     * Prints all the objects in the given array handling nested collections/arrays as records.
     *
     * <p>
     * If the given array only contains simple objects, this method will print a single record like
     * {@link #printRecord(Object...)}. If the given collections contains nested collections/arrays those nested
     * elements will each be printed as records using {@link #printRecord(Object...)}.
     * </p>
     *
     * <p>
     * Given the following data structure:
     * </p>
     *
     * <pre>
     * <code>
     * String[][] data = new String[3][]
     * data[0] = String[]{ "A", "B", "C" };
     * data[1] = new String[]{ "1", "2", "3" };
     * data[2] = new String[]{ "A1", "B2", "C3" };
     * </code>
     * </pre>
     *
     * <p>
     * Calling this method will print:
     * </p>
     *
     * <pre>
     * <code>
     * A, B, C
     * 1, 2, 3
     * A1, B2, C3
     * </code>
     * </pre>
     *
     * @param values
     *            the values to print.
     * @throws IOException
     *             If an I/O error occurs
     */
    public void printRecords(final Object... values) throws IOException {
        obj.invokeMember("print_records", values);
    }

    /**
     * Prints all the objects in the given JDBC result set.
     *
     * @param resultSet
     *            result set the values to print.
     * @throws IOException
     *             If an I/O error occurs
     * @throws SQLException
     *             if a database access error occurs
     */
    public void printRecords(final ResultSet resultSet) throws SQLException, IOException {
        obj.invokeMember("print_records", resultSet);
    }
}
